### ISSUE TITLE
Example issue, or llvm-mos-sdk bug, or simple configuration error?

### ISSUE THREAD

**Issue by 3vi1** [rolandshacks/vs64#77]
I was looking at the examples, and the cpp example doesn't seem to generate the "Hello World" behavior I would expect from the code. It just kind of switches character sets and crashes.

[![Odd behavior](https://i9.ytimg.com/vi/G5eCKWAik_o/mqdefault.jpg?sqp=CNzh27gG-oaymwEmCMACELQB8quKqQMa8AEB-AH-CYAC0AWKAgwIABABGCkgSyh_MA8=&rs=AOn4CLAG17BpHvEGFDS1521XyLvXq6FzyQ)](https://youtu.be/G5eCKWAik_o "Odd behavior")

Trying to single-step through it ends up with it just cycling through testfunc() over and over.

This is using the current git version of llvm-mos-sdk compiled from source with VICE 3.7.1 (Ubuntu package 3.7.1+dfsg1-3) on Linux:

```
(base) 19:16:48 evil@H510 /opt/llvm-mos-sdk/bin» ./mos-c64-clang++ --version
clang version 20.0.0git (https://github.com/llvm-mos/llvm-mos 8445a95309f2e8673c967f78739e18f32f5208bf)
Target: mos-c64
Thread model: posix
InstalledDir: /opt/llvm-mos-sdk/bin
Configuration file: /opt/llvm-mos-sdk/bin/mos-c64.cfg
```

I deleted the build directory and let it rebuild everything from scratch with the same results.

Is there maybe just something simple I'm just overlooking?

**Comment by rolandshacks** [comment]
Great find!
Two things:
- With v18, LLVM-MOS changed the behavior to not exit from main, instead loop infinetely (that's because they heavily mess up the zero page and cannot restore)
- For whatever reason, the very last line of that inline assembly file with a 'ret' is missing (my fault, ...)

Correct assembler.asm file:
```
;
; Assembler Code
;

.global testfunc

testfunc:
    pha
    lda #$aa
    pla
    rts
```

Better main.cpp:
```
//
// Hello, world!
//

#include <stdio.h>

#include "src/include.h"

extern "C" void testfunc(void);
extern const unsigned char data[];

int main(void) {
    testfunc();
    printf("%s\n", data);
    for (;;) *((char*) 53280) += 1;
    return 0;
}
```

And last but not least, printf will work a lot better if the RAW data file ends with a zero-byte (because for simple demonstration, I just utilize the raw data as string, which the resource compiler will not automatically terminate with a \0).

I will fix the example and will release a new version of VS64 asap.

Thanks for reporting, it is really appreciated!

Roland

**Comment by rolandshacks** [comment]
PS. Btw - you can open the build folder and then just click on the generated PRG. That will open a disassembly in vscode, helps a lot to find these little issues and to learn what LLVM actually produces...

**Comment by 3vi1** [comment]
> PS. Btw - you can open the build folder and then just click on the generated PRG. That will open a disassembly in vscode, helps a lot to find these little issues and to learn what LLVM actually produces...

Oh, I tried that... but llvm added a *lot* of code, and my 6510 expertise is about 30 years rusty.  :)  I've definitely got some cobwebs to shake off.

Thank you for the quick and helpful explanation of what was going on!  I look forward to playing with VS64 some more this weekend; this is a really cool extension.

**Comment by rolandshacks** [comment]
Guess we're in the same boat. But it's truly amazing to have access to all documentation, books, etc. - which I could not afford as a German teenager at that time. The idea of VS64 was born while working on autonomous systems, built with 50k compute devices with most powerful GPUs and FPGAs, complemented by another 200k R&D sensors, equipment, storing petabytes of data in the cloud, training AI models while paying a fortune to AWS, all desks with 2x 4k Monitors, everyone had a workstation, a variety of modern IDEs and compilers, etc. And still ... there was something in the early days that you cannot get anymore these days. ... and the question was: what if I throw all kinds of modern things together to develop for the C64 as I would do for a modern target? That's it, let's see how far this can get :-)

**Comment by 3vi1** [comment]
Totally agree!  I spent a lot of fun hours manually disassembling the data statements from programs in Compute!'s Gazette back in those days.  :)

The resource compiler is definitely one of the next things I want to check out.  I had played with making CBM Prg Studio work on Linux via WINE because of its built-in resource features... but I'm not crazy about the editor itself because it's closed source, not supported natively on Linux, and we can't easily contribute bug-fixes.

I've played with various other cross-platform tools that might be nice integrate as well - I just need to see if you haven't already made them unnecessary.  :)

I tested your recommended changes to the cpp demo, and they work perfectly.  I created a fork and packaged them up in a pull request to save you the time.  If I can do anything else to help, I'll be sure to send more requests your way!

Thanks again.
